### ISSUE TITLE
[no-hooks] breaking change introduced in `6.2.2`.

### ISSUE THREAD

**Issue by BBosman** [lo1tuma/eslint-plugin-mocha#225]
After updating from `6.2.1` to `6.2.2` I start receiving unexpected violations on `mocha/no-hooks` without any code being changed on my side.

**Repro**

```JSON
{
  "rules": {
    "mocha/no-hooks": "error"
  }
}
```

```TS
import { Aurelia } from '@aurelia/runtime';
import { assert, TestContext } from '@aurelia/testing';

describe('Repro', function () {
  async function setup() {
    const ctx = TestContext.createHTMLTestContext();
    const host = ctx.doc.createElement('div');
    const au = ctx.wnd['au'] = new Aurelia(ctx.container).app({ host: host });

    await au.start().wait();

    return host;
  }
  it('loads viewports left and right', async function () {
    const host = await setup(); // <-- Unexpected use of Mocha `setup` hook - eslint(mocha/no-hooks)

    assert.includes(host.textContent, 'left', `host.textContent`);
    assert.includes(host.textContent, 'right', `host.textContent`);
  });
});
```

**Expected Result**

For this to not start failing all of a sudden.

**Actual Result**

I get `Unexpected use of Mocha `setup` hook - eslint(mocha/no-hooks)` on the indicated line.

** Additional Info ** 

If I rename `setup` to something else the violation isn't triggered. But as this would need thousands of changes throughout our codebase I'm not keen on having to do that right now.

My suggestion would be to undo the introduced TDD support from #220 and release that as `6.2.3` and release a new `7.0.0` with the TDD support.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `eslint-plugin-mocha` | `6.2.2` |
| `TypeScript`                       | `3.7.3` |
| `ESLint`                           | `6.7.2` |
| `node`                             | `12.12.0` |
| `npm`                              | `6.13.4` |

**Comment by lo1tuma** [comment]
This wasn’t a breaking change. Not supporting the TDD interface was a bug in the `no-hooks` rule, see #219.
What your example code does is it shadows the global `setup` function that mocha already provides. Currently the `no-setup` rule just looks at the name of function calls in order to determine whether it is a mocha hook or not. We could improve that by using scope analysis und only trigger when the referenced function in a function call comes either from the global scope, from an import statement or a `require` call.

**Comment by fkleuver** [comment]
Ah, we weren't aware of the `setup` hook actually. Probably better if we use a different name then because we'd rather avoid shadowing stuff.

Scope analysis might incur a perf penalty (linting is already pretty slow as it is) for something that could be solved with a simple exclusion rule. Is there an option to specify which particular hooks to include/exclude?

**Comment by lo1tuma** [comment]
> Ah, we weren't aware of the `setup` hook actually. Probably better if we use a different name then because we'd rather avoid shadowing stuff.

Yeah, I can recommend the `no-shadow` core rule from ESLint. When configured correctly (i.e. using `env: [ 'mocha' ]`) it would already complain about shadowing globals.

> Scope analysis might incur a perf penalty

[ESlint does it anyway](https://eslint.org/docs/developer-guide/working-with-rules#context-getscope), so I don’t think that using this data in a rule would have a significant performance impact. But I might be wrong 😄.

> Is there an option to specify which particular hooks to include/exclude?

No, there is only an option to add additional/unofficial mocha function names that might be added by mocha extensions or wrappers.

**Comment by BBosman** [comment]
Closing this as it's indeed expected behaviour based on your extra information.

We've renamed our `setup` functions (which took less work than expected) so we resolved the issue on our side as well.

Thanks for looking into it. 👍
